### PR TITLE
[Merged by Bors] - feat: MeasureTheory.Integral.Bochner: integral_fintype and similar

### PR DESCRIPTION
**Pull request by nomeata**
This adds some lemmas about `integral` where we already have the corresponding
lemmas for `lintegral`. The goal is `integral_fintype`, which rewrites an
integral over a finite type as a finite sum over the elements, with singleton
measures (these singleton measures can then further be simplified when the
measure comes from a `Pmf`, but that will follow some other time.).

Also fixes lemma name `integral_eq_lintegral_pos_part_sub_lintegral_neg_part` in
the module comments.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Integral/Bochner.lean [hunk 1]
```diff
@@ -1776,6 +1782,59 @@ theorem integral_mul_le_Lp_mul_Lq_of_nonneg {p q : ℝ} (hpq : p.IsConjugateExpo
 set_option linter.uppercaseLean3 false in
 #align measure_theory.integral_mul_le_Lp_mul_Lq_of_nonneg MeasureTheory.integral_mul_le_Lp_mul_Lq_of_nonneg
 
+theorem integral_countable' [Countable α] [MeasurableSingletonClass α] {μ : Measure α} {f : α → ℝ}
+  (hf : Integrable f μ):
+  ∫ a, f a ∂μ = ∑' a, f a * (μ {a}).toReal := by
```
Comment (nomeata) [inline Mathlib/MeasureTheory/Integral/Bochner.lean:1787]: I wonder if this should be `(hf : Integrable f (μ.restrict s))` (aka `IntegrableOn`, but that notation isn’t available yet) instead, for a weaker precondition.

Maybe it helps to have something like
```
/-- If `s` is a countable set, then `μ.restrict s` is a sum of Dirac measures. -/
theorem restrict_eq_sum {s : Set α} (hs : Countable s) [MeasurableSingletonClass α] (μ : Measure α)
  : μ.restrict s = sum (s.indicator fun a => μ {a} • dirac a)
```
to match [`map_eq_sum`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Dirac.html#MeasureTheory.Measure.map_eq_sum), and then not go via the subtypes, but I am uncertain. I probably needs expert guidance.

File: Mathlib/MeasureTheory/Integral/Bochner.lean [hunk 2]
```diff
@@ -1776,6 +1782,59 @@ theorem integral_mul_le_Lp_mul_Lq_of_nonneg {p q : ℝ} (hpq : p.IsConjugateExpo
...
+     mul_comm]
+
+theorem integral_singleton [MeasurableSingletonClass α] {μ : Measure α} (f : α → ℝ) (a : α):
+  ∫ a in {a}, f a ∂μ = f a * (μ {a}).toReal := by
+  simp only [Measure.restrict_singleton, integral_smul_measure, integral_dirac, smul_eq_mul,
+     mul_comm]
+
+theorem integral_countable [MeasurableSingletonClass α] (f : α → ℝ)
+  {s : Set α} (hs : s.Countable) (hf : Integrable f (μ.restrict s)) :
+  ∫ a in s, f a ∂μ = ∑' a : s, f a * (μ {(a : α)}).toReal := by
```
Comment (nomeata) [inline Mathlib/MeasureTheory/Integral/Bochner.lean:1805]: I wonder if this should be `(hf : Integrable f (μ.restrict s))` (aka `IntegrableOn`, but that notation isn’t available yet) instead, for a weaker precondition.

Maybe it helps to have something like
```
/-- If `s` is a countable set, then `μ.restrict s` is a sum of Dirac measures. -/
theorem restrict_eq_sum {s : Set α} (hs : Countable s) [MeasurableSingletonClass α] (μ : Measure α)
  : μ.restrict s = sum (s.indicator fun a => μ {a} • dirac a)
```
to match [`map_eq_sum`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Dirac.html#MeasureTheory.Measure.map_eq_sum), and then not go via the subtypes, but I am uncertain. I probably needs expert guidance.

**Comment by nomeata** [comment]
The motivation of the present PR is to be able to do https://github.com/leanprover-community/mathlib4/pull/6454

**Comment by nomeata** [comment]
I’ll be traveling without a laptop for week, and will address review comments afterwards. Should anyone want to adopt this PR and see it through, that’s fine with me as well.

File: Mathlib/MeasureTheory/Integral/Bochner.lean [hunk 3]
```diff
@@ -1776,6 +1782,56 @@ theorem integral_mul_le_Lp_mul_Lq_of_nonneg {p q : ℝ} (hpq : p.IsConjugateExpo
...
+  congr 1 with a : 1
+  rw [integral_smul_measure, integral_dirac, mul_comm, smul_eq_mul, Measure.sum_smul_dirac]
+
+theorem integral_singleton' {μ : Measure α} {f : α → ℝ} (hf : StronglyMeasurable f) (a : α):
+  ∫ a in {a}, f a ∂μ = f a * (μ {a}).toReal := by
+  simp only [Measure.restrict_singleton, integral_smul_measure, integral_dirac' f a hf, smul_eq_mul,
+     mul_comm]
+
+theorem integral_singleton [MeasurableSingletonClass α] {μ : Measure α} (f : α → ℝ) (a : α):
+  ∫ a in {a}, f a ∂μ = f a * (μ {a}).toReal := by
```
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Integral/Bochner.lean:1799]:
```suggestion
theorem integral_singleton [MeasurableSingletonClass α] {μ : Measure α} (f : α → ℝ) (a : α) :
    ∫ a in {a}, f a ∂μ = f a * (μ {a}).toReal := by
```

File: Mathlib/MeasureTheory/Integral/Bochner.lean [hunk 4]
```diff
@@ -1632,10 +1633,15 @@ theorem MeasurePreserving.integral_comp {β} {_ : MeasurableSpace β} {f : α 
   h₁.map_eq ▸ (h₂.integral_map g).symm
 #align measure_theory.measure_preserving.integral_comp MeasureTheory.MeasurePreserving.integral_comp
 
-theorem set_integral_eq_subtype {α} [MeasureSpace α] {s : Set α} (hs : MeasurableSet s)
-    (f : α → G) : ∫ x in s, f x = ∫ x : s, f x := by
+theorem set_integral_eq_subtype' {α} [MeasurableSpace α] {μ : Measure α} {s : Set α}
+  (hs : MeasurableSet s)
+    (f : α → G) : ∫ x in s, f x ∂μ =
```
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Integral/Bochner.lean:1638]:
```suggestion
    (hs : MeasurableSet s) (f : α → G) : 
    ∫ x in s, f x ∂μ =
```

**Review by RemyDegenne — COMMENTED** [review]
Thanks!
I left a few comments about code style. I'll come back to this for a full review later.

**Comment by RemyDegenne** [comment]
Usually I would simply have seen that you had pushed modifications to the branch and I would have come back to this PR without you needing to ping me, but I was on vacation recently and I lost track of those reviews. Sorry about the long delay. I'll review it shortly.

File: Mathlib/MeasureTheory/Integral/Bochner.lean [hunk 5]
```diff
@@ -1776,6 +1782,58 @@ theorem integral_mul_le_Lp_mul_Lq_of_nonneg {p q : ℝ} (hpq : p.IsConjugateExpo
...
+theorem integral_countable [MeasurableSingletonClass α] (f : α → ℝ)
+    {s : Set α} (hs : s.Countable) (hf : Integrable f (μ.restrict s)) :
+    ∫ a in s, f a ∂μ = ∑' a : s, f a * (μ {(a : α)}).toReal := by
+  have hi : Countable { x // x ∈ s } := Iff.mpr countable_coe_iff hs
+  have hf' : Integrable (fun (x : s) => f x) (Measure.comap Subtype.val μ) := by
+    rw [← map_comap_subtype_coe, integrable_map_measure] at hf
+    apply hf
+    exact Integrable.aestronglyMeasurable hf
+    exact Measurable.aemeasurable measurable_subtype_coe
+    exact Countable.measurableSet hs
```
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Integral/Bochner.lean:1812]: Please use dots to focus on each of the three goals. It makes the structure of the proof clearer.
```suggestion
    · exact Integrable.aestronglyMeasurable hf
    · exact Measurable.aemeasurable measurable_subtype_coe
    · exact Countable.measurableSet hs
```

**Review by RemyDegenne — APPROVED** [review]
It looks good!
bors r+

**Comment by RemyDegenne** [comment]
Sorry, typo:
bors r-
bors d+

**Pull request closed without merging**